Add harmonic keywheel ui

poelzi opened the pull request:
Add a Dialog with the harmonic keywheel as a cheat sheet.
Clicking on the wheel will cycle through the different notations.

This is very useful on B2B cases when both DJs are using different notations.
Also it helps when using pitching to lookup the next key.

TODO:

- [ ] Some of the styling I used in the SVG is not supported by the qt svg support. Would be good if someone from the skinning team would give it some polishing.
- [ ] Maybe fix the aspect ratio but does not work

poelzi committed:
Add harmonic keywheel ui

Add a Dialog with the harmonic keywheel as a cheat sheet.
Clicking on the wheel will cycle through the different notations.

This is very useful on B2B cases when both DJs are using different notations.
Also it helps when using pitching to lookup the next key.

Be-ing commented:
Neat idea. Can you post a screenshot?

Be-ing commented:
It looks odd to me to put F at the top. Every other diagram of the circle of fifths I have seen puts C at the top.

poelzi committed:
Update keywheel.

- white text for notation of traditional keys
- move C to top position
- rotate key notation to make it more compact on some keys

poelzi committed:
better keyboard mouse event handling in keywheel window

* cycle through in both directions with mouse
* cycle with keyboard TAB/Shift+TAB

remove debug output

poelzi commented:
![selection_002](https://user-images.githubusercontent.com/66107/40839371-8e918b8a-65a2-11e8-9c4f-da58b5645da8.png)

Updated version. TAB + Shift TAB now also works

Be-ing commented:
I don't think the window needs its own Close button. There is already the window manager's close button.

poelzi commented:
Not all window managers have a close button. Awesome for example does have one. Dialogs without close buttons are strange even stranger. It's existence also ensures the window handles key events correctly.

Be-ing commented:
You could make the dialog close when clicking on it like the cover art dialog.

> It's existence also ensures the window handles key events correctly.

What key events would this simple dialog need to handle?

esbrandt commented:
Thanks for working on this feature.

I had not the chance to test the live code ( can´t build mixx with qt5 yet), so for now just some comments from looking at the screenshoot.

* For Open Key Notation, we should not try to establish our own color set in the key wheel, lets use the established colors instead.

  The color palette below is derived from the Open Key Notation, and already used in Traktor,  Scratch Pro  etc...  
  The palette might be useful later , when displaying keys in the library in different colors, or when displaying key changes in the waveform.
 Just save as ``*.gpl`` for use with Inkscape / Gimp. ``*.ase`` palette available upon request.

```txt
GIMP Palette
Name: Open Key Notation colors
Columns: 1
# Created for <https://mixxx.org/>
# Based upon Open Key Notation
# <https://www.beatunes.com/en/open-key-notation.html>
# licensed under CC0 1.0 Universal (CC0 1.0)
# <https://creativecommons.org/publicdomain/zero/1.0/>

252  73  73	1d/1m #FC4949
254 100  45	2d/2m #FE642D
249 140  39	3d/3m #F98C27
254 214   0	4d/4m #FED600
153 254   0	5d/5m #99FE00
 66 254  62	6d/6m #42FE3E
 10 213 143	7d/7m #0AD58F
 10 231 231	8d/8m #0AE7E7
  4 201 254	9d/9m #04C9FE
 61 138 253	10d/10m #3D8AFD
172 100 254	11d/11m #AC64FE
253  63 234	12d/12m #FD3FEA

```
  ![gimp palette editor-1](https://user-images.githubusercontent.com/4525897/40885648-f61aa5b4-672a-11e8-8538-be6f2673b8e6.png)

 * Should the modal key widget really overlay the application widget? We only allow it for the preferences, dev console, and properties editor - all of them usually not open during performance.
I understand the keywheel is meant to be a cheat sheet, but someone might have it open pretty often, if mixing by key ( and Mixxx not having a related-key feature yet). It has the potential to grow into a even more powerful feature.

daschuer commented:
It can be a feature in the new three column library.It was designed exactly for such additional infos.

Be-ing commented:
I think we should be removing items from the menu bar so we can remove the menu bar entirely. Perhaps the place to this image would be somewhere in the library, but I don't want to make this little branch depend on the library redesign branch. So I think adding this design debt is okay for now.

Be-ing commented:
There are a few oddities with traditional notation. The keys are displayed twice per sector. Also, D sharp minor/E flat minor does not fit in the allocated space.
![screenshot from 2018-06-11 16-14-58](https://user-images.githubusercontent.com/9455094/41257517-bfd0d32a-6d92-11e8-8e70-f5f05124384a.png)

Be-ing commented:
> I don't think the window needs its own Close button. There is already the window manager's close button.

I didn't realize that the dialog responds to click events already. So I think the Close button is okay.

Be-ing commented:
@poelzi could you take care of the above issues in the next week? If so I think we could merge this for 2.2.

poelzi committed:
Update keywheel colors to match openkey defined colors

poelzi committed:
Skip traditional mode as it only shows redundent informations.

Change colors of wheel to be more readable.

poelzi commented:
Did not know there were official color wheel colors, thanks. I changed the svg to match those.

@Be-ing The text is a simple replacement of the body in the template. I removed the traditional display so only on the custom setting something strange could be displayed. I don't see a way to easily ensure the text will fit the box.

Be-ing commented:
I'm confused. The last commit doesn't seem to change the behavior for me. However, I don't think it's a good approach. I want to see the diagram with traditional notation -- I just don't want to see all the keys listed multiple times in each sector. Thinking about it more, I'm a bit confused why this feature is even helpful with the numbered key notations because it is so easy to think of how keys are related with those notations.

poelzi commented:
@Be-ing I think your custom settings are like traditional ?
Of coures custom is always showed, I'm not testing if the settings are identical to traditional and in this case I'm skipping custom as well.

The reason for having also numbered wheel on it is to look up for example openkey or lancelot if you need it. Also calculating the key pitch up/down will be in is not something I'm always 100% certain of while 140 bpm is hammering on your head.

Ok, so, you want a extra layout just for traditional view without the static information used for orientation ?

Be-ing commented:
> @Be-ing I think your custom settings are like traditional ?

No, I checked that I am not using a custom notation.

> Ok, so, you want a extra layout just for traditional view without the static information used for orientation ?

Yes.

poelzi committed:
static cast

Holzhaus commented:
Could this be used as editor for a `KeyDelegate` that opens when trying to edit the Key column in the Track table view?

uklotzde committed:
Merge branch '2.3' of git@github.com:mixxxdj/mixxx.git

uklotzde committed:
Merge branch 'main'

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git

daschuer committed:
Merge remote-tracking branch 'upstream/main'

poelzi committed:
Add keywheel to cmake build

poelzi committed:
Bring keywheel uptodate

Fix crash when loading.
Skip notations with redundent informations
Try to break long lines

daschuer committed:
Merge remote-tracking branch 'upstream/main'

daschuer committed:
Merge remote-tracking branch 'upstream/main'

poelzi committed:
Move keywheel to help menu.

Change shortcut to F2 as it is close to F1 for help.

poelzi committed:
Use new connect syntax

poelzi committed:
fix styling

poelzi committed:
Show traditional layer but hide legend instead.

daschuer reviewed (changes requested):
Thank you for this nice addition. 
Here some comments.

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into harmonic-wheel

poelzi committed:
Simplify notation update code

poelzi committed:
Use ControlObject to recieve settings

poelzi committed:
Remove unused variable

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into harmonic-wheel

poelzi commented:
@Holzhaus I tried the other day to add user interaction but qtsvg is just to limited. Maybe later I will replace the svg with a custom widget and make the segments selectable. I would like to make the segments selectable in the dialog and see which key will it transform into when pitched up/down. But not now, I have more important things to implement :)
@be-ing the legend is now hidden on traditional key display

poelzi committed:
fix windows casting warning

poelzi commented:
ready to merge

poelzi commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,148 @@
+    m_notation = newNotation;
+    // we update the SVG nodes with the new value
+    updateDisplay();
+}
```
fixt

poelzi committed:
Merge remote-tracking branch 'origin/main' into main

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into main

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into main

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into main

poelzi commented:
@daschuer ping

poelzi commented:
@Be-ing wanna merge ? I addressed your duplicated information regards and rebased to main.

Be-ing commented:
There are some conflicts now.

Be-ing commented:
I'm not seeing a way to show the dialog? It should be in the View menu, right?

poelzi commented:
@Be-ing i moved it to help, since it makes more sense there. Shortcut F2

Be-ing commented:
D sharp minor/E flat minor label goes outside its sector:
![image](https://user-images.githubusercontent.com/9455094/102398765-526d9880-3fa5-11eb-9a27-e6725f55fadc.png)

Be-ing commented:
This view shows redundant labels:
![image](https://user-images.githubusercontent.com/9455094/102398941-8e086280-3fa5-11eb-9852-26d6e3e01111.png)

daschuer commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,153 @@
+        KeyUtils::KeyNotation::LancelotAndTraditional};
+} // namespace
+
+DlgKeywheel::DlgKeywheel(QWidget* parent, UserSettingsPointer pConfig)
```
```suggestion
DlgKeywheel::DlgKeywheel(QWidget* parent, const UserSettingsPointer& pConfig)
```

daschuer commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,153 @@
+    ui->setupUi(this);
+    QDir resourceDir(m_pConfig->getResourcePath());
+    auto svgPath = resourceDir.filePath(kKeywheelSVG);
+    qDebug() << svgPath;
```
I think that should be commented out.

daschuer commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,153 @@
+        return true;
+    }
+    // standard event processing
+    return QObject::eventFilter(obj, event);
```
I think this should be
```suggestion
    return QDialog::eventFilter(obj, event);
```
..,to not jump the inheritance.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -519,6 +517,20 @@ void WMainMenuBar::initialize() {
     QString externalLinkSuffix = " =>";
 
+    QString keywheelTitle = tr("Show Keywheel");
+    QString keywheelText = tr("Show keywheel");
```
This is a translator trap, because they cannot know which string is displayed where.
Please add transla tor comment. //:

daschuer reviewed (changes requested):
I have left some more comments, sorry for the delay.

I think for this PR the menu entry is OK. I think the use case is that user open this window once and than move and minimize it using the windows controls. 

I need to test if this really works flawlessly.

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into harmonic-wheel

poelzi committed:
Cleanup keywheel code

poelzi commented:
@Be-ing I found no way to fix the size problems on long text with the SVG implemention of Qt. The SVGTiny implementation is very limited :(

Be-ing commented:
Okay, thank you for trying.

Be-ing commented:
clazy failed on Qt's autogenerated ui header??

poelzi commented:
Looks like it. Im changing it to a normal button now.

poelzi committed:
Fix clazy problem with buttonbox

daschuer commented:
A conflict has developed. Please merge upstream/main and resolve the conflict.

poelzi committed:
Merge branch 'main' into harmonic-wheel

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into harmonic-wheel

poelzi commented:
@daschuer fixed

daschuer commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,158 @@
+    }
+    m_domDocument.setContent(&xmlFile);
+
+    // FIXME(poelzi): try to prevent aspect ratio. This fails unfortunatelly
```
Does this comment still apply?

daschuer commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,158 @@
+    QDomElement domElement;
+
+    QDomNodeList node_list = m_domDocument.elementsByTagName(QStringLiteral("text"));
+    for (int i = 0; i < node_list.count(); i++) {
```
Can this become a range based loop?

poelzi replied:
no, the QDomNodeList API is very rudimentary.

daschuer commented on src/dialog/dlgkeywheel.h:
```diff
@@ -0,0 +1,34 @@
+    bool isHiddenNotation(KeyUtils::KeyNotation notation);
+    KeyUtils::KeyNotation m_notation;
+    QDomDocument m_domDocument;
+    Ui::DlgKeywheel* ui;
```
This can be removed, because Ui::DlgKeywheel is the parent class

daschuer commented on src/mixxx.cpp:
```diff
@@ -924,6 +932,23 @@ void MixxxMainWindow::slotHelpAbout() {
 
+void MixxxMainWindow::slotShowKeywheel(bool toggle) {
+    if (m_pKeywheel == nullptr) {
+        m_pKeywheel = new DlgKeywheel(this, m_pCoreServices->getSettings());
```
you can use here parented_ptr to verify that DlgKeywheel is controlled by the Qt object tree.

daschuer reviewed (changes requested):
I don't see the View menu entry, F2 works. 

I have left some more comments inline.

poelzi committed:
Cleanup Keywheel code

poelzi commented on src/dialog/dlgkeywheel.cpp:
```diff
@@ -0,0 +1,158 @@
+
+    m_domDocument.save(stream, QDomNode::EncodingFromDocument);
+
+    ui->graphic->load(str.toUtf8());
```
no, because a QString means load the file and we have the content here. I added a comment

poelzi commented:
@daschuer I moved the entry into the help menu, because it is more of a help then some feature that is switched in the view menu.

daschuer commented:
The wheel still starts as flat. Once you grep the edge of it, it snaps into a round shape. I assume this happens, because the svg is loaded after the size is set. Maybe it is fixed if you swap it?

daschuer commented:
Or maybe it helps to call resizeEvent() manually ...?

poelzi commented:
@daschuer I tried that, does not work. I tried calling it on show(), after initialzation. I tried to actually resize the window on first show but this behaviour persists. I also don't understand why the left spacer only works when the window is very large, but not when the window is small. I think the SVG widget is just buggy in multiple places.
I consider this: as good as it gets without rewriting the complete thing or someone with better QT understanding fixes it.

poelzi committed:
Add workaround for wrong SVG size after loading

poelzi commented:
I found one combination for at least 5.15 that seems to work

daschuer reviewed (approved):
LGTM, Thank you for this nice addition. 
Hopefully the users find the new menu entry.

ronso0 commented:
we could put a color wheel icon in the tollbars so this is quickly available, also once we have switched to the hideable toolbar.

poelzi commented:
If someone has time and willingness, the SVG could be replaced by a custom widget that draws the cicle and text. This way, it could become clickable and we could use it as a delegate in the key column. I found no way todetect which element was clicked.

esbrandt commented:
The keywheel widget is displayed blank. 
`images/keywheel/keywheel.svg` gets called,
but is missing from https://github.com/mixxxdj/mixxx/blob/main/res/mixxx.qrc

tested with 2.4.0-alpha-pre (build main r7757) macOS

esbrandt commented on src/widget/wmainmenubar.cpp:
```diff
@@ -532,6 +530,22 @@ void WMainMenuBar::initialize() {
+    m_pViewKeywheel->setShortcut(
+            QKeySequence(m_pKbdConfig->getValue(
+                    ConfigKey("[KeyboardShortcuts]", "ViewMenu_ShowKeywheel"),
+                    tr("F2", "Menubar|View|Show Keywheel"))));
```
This shortcut clashes with the existing shortcut for `[Channel1],rate_perm_up`, see https://github.com/mixxxdj/mixxx/blob/main/res/keyboard/en_US.kbd.cfg#L43

poelzi commented:
thanks @esbrandt , fixes in pr
